File: kattis/medium/teque.py
import sys
from collections import deque
from numba import jit

class Teque:
    def __init__(self):
        self.front = deque()
        self.back = deque()

    @jit(nopython=True)
    def push_front(self, x):
        self.front.appendleft(x)
        self.balance()

    @jit(nopython=True)
    def push_back(self, x):
        self.back.append(x)
        self.balance()

    @jit(nopython=True)
    def push_middle(self, x):
        self.front.append(x)
        self.balance()

    @jit(nopython=True)
    def get(self, index):
        return self.front[index] if index < len(self.front) else self.back[index - len(self.front)]

    @jit(nopython=True)
    def balance(self):
        while len(self.front) > len(self.back) + 1:
            self.back.appendleft(self.front.pop())
        while len(self.back) > len(self.front):
            self.front.append(self.back.popleft())

def main():
    lines = int(sys.stdin.readline())
    teque = Teque()

    for _ in range(lines):
        action, arg = sys.stdin.readline().split()

        if action == "push_front":
            teque.push_front(int(arg))
        elif action == "push_back":
            teque.push_back(int(arg))
        elif action == "push_middle":
            teque.push_middle(int(arg))
        elif action == "get":
            sys.stdout.write(str(teque.get(int(arg))) + '\n')

if __name__ == "__main__":
    main()

File: kattis/medium/cpp/teque.cpp
//
// Created by BBJ on 02.01.2024.
//
#include <iostream>
#include <deque>
using std::cin;
using std::cout;
using std::string;
using std::getline;
using std::endl;

template <typename T>
class teque {
public:
    std::deque<T> front;
    std::deque<T> back;

    void push_front(T x) {
        front.push_front(x);
        balance();
    }

    void push_back(T x) {
        back.push_back(x);
        balance();
    }

    void push_middle(T x) {
        front.push_back(x);
        balance();
    }

    T get(int index) {
        return index < front.size() ? front[index] : back[index - front.size()];
    }

    void balance() {
        while (front.size() > back.size()  + 1) {
            back.push_front(front.back());
            front.pop_back();
        }
        while (back.size() > front.size()) {
            front.push_back(back.front());
            back.pop_front();
        }
    }
};

int main() {
    int lines; cin >> lines;
    teque<int> teque;

    for (int i=0; i < lines; i++) {
        string action;
        int arg;
        cin >> action >> arg;

        if (action == "push_front")
            teque.push_front(arg);
        else if (action == "push_back")
            teque.push_back(arg);
        else if (action == "push_middle")
            teque.push_middle(arg);
        else if (action == "get")
            cout << teque.get(arg) << endl;
    }
    return 0;
}